compiler/rustc_lint_defs/src/builtin.rs

```diff
@@ -2965,6 +2965,42 @@ declare_lint! {
     };
 }
 
+declare_lint! {
+    /// The `ptr_to_int_cast` lint checks casting pointers to integers other than `usize`. These
+    /// casts are often bugs.
+    ///
+    /// ### Example
+    ///
+    /// ```rust,compile_fail
+    /// #![deny(ptr_to_int_cast)]
+    ///
+    /// fn main() {
+    ///     let x = 100_000_000;
+    ///     let y = u16::max as u32; // a function pointer, cast to u32, causing an overflow
+    ///     println!("{}", x > y); // prints `false` instead of `true`
+    /// }
+    /// ```
+    ///
+    /// {{produces}}
+    ///
+    /// ### Explanation
+    ///
+    /// The example above shows accidentally casting a function (rather than an integer constant)
+    /// to an integer type smaller than the pointer size. The cast overflows and the program yields
+    /// incorrect result.
+    ///
+    /// Casting pointers to integer types other than `usize` are often bugs. This lint warns
+    /// against those cases.
+    ///
+    /// (In the future we may want to disallow this case entirely)
+    ///
+    /// To cast a pointer to e.g. u32 without triggering the lint, first cast to usize:
+    /// `ptr as usize as u32`.
+    pub PTR_TO_INT_CAST,
+    Allow,
```

Comment:
> Note that the lint is not enabled by default, I'm guessing we want to enable it by default before shipping, right?

I think it would be reasonable to make this `Warn` by default. We'll need to get the lang team to sign off on the new lint.